extensions/async-extensions/src/commonMain/kotlin/app/cash/sqldelight/async/coroutines/QueryExtensions.kt

```diff
@@ -1,11 +1,24 @@
 package app.cash.sqldelight.async.coroutines
 
 import app.cash.sqldelight.ExecutableQuery
+import app.cash.sqldelight.db.QueryResult
 
 suspend fun <T : Any> ExecutableQuery<T>.awaitAsList(): List<T> = execute { cursor ->
+  val first = cursor.next()
   val result = mutableListOf<T>()
-  while (cursor.next()) result.add(mapper(cursor))
-  result
+
+  // If the cursor isn't async, we want to preserve the blocking semantics and execute it synchronously
+  if (first is QueryResult.AsyncValue) {
+    QueryResult.AsyncValue {
+      if (first.await()) result.add(mapper(cursor)) else return@AsyncValue result
+      while (cursor.next().value) result.add(mapper(cursor))
```

Comment:
Don't we need to call `await` for this `next` too? Assuming all next calls of the same cursor are always async (or sync).

Reply:
Oh yes, good catch.